subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/VariantResolveMetadata.java

```diff
@@ -46,5 +46,11 @@ public interface VariantResolveMetadata extends VariantGraphResolveMetadata.Subv
     boolean isExternalVariant();
 
     interface Identifier {
+        /**
+         * Is the variant identified by this identifier eligible for caching?
+         *
+         * Only variants from a project component are eligible for caching.
+         */
+        boolean isEligibleForCaching();
```

Comment:
It seems odd for this to live on the `Identifier` and not the actual variant being identified. I am not really sure _why_ only project component variants are eligible for caching, but regardless, it seems this method only returns true for `ComponentConfigurationIdentifier` instances when the parent component is a `ProjectComponentIdentifier` instance. (and delegating implementations).

What do you think about moving this method to `VariantResolveMetadata` instead of `VariantResolveMetadata.Identifier`? This would mean we'd update `LocalVariantMetadata`s created as a result of `DefaultLocalComponentMetadataBuilder#addConfiguration`. I've prepared a commit which makes this change if you choose to implement it: 0e380861a8e559ee9df135d7aaf9068241cf2edb

Reply:
The proposal makes sense to me.

As for why we only cache project variants, that's the result of doing a fix focused on the bigger problem: explosion of variants from projects in large builds.

Also, third party components can have their variants mutated by component metadata rules, which makes caching them centrally just a bit trickier.